dali/pipeline/operators/transpose/transpose.h

```diff
@@ -24,36 +24,66 @@
 
 namespace dali {
 
+namespace detail {
+
+template <typename T>
+T Permute(const T& in, const std::vector<int>& permutation) {
+  T out = in;
+  for (size_t i = 0; i < permutation.size(); i++) {
+    auto idx = permutation[i];
+    out[i] = in[idx];
+  }
+  return out;
+}
+
+}  // namespace detail
+
+
 template <typename Backend>
 class Transpose : public Operator<Backend> {
  public:
-  explicit inline Transpose(const OpSpec &spec) :
-    Operator<Backend>(spec),
-    perm_(spec.GetRepeatedArgument<int>("perm")) {
-      DALI_ENFORCE([](std::vector<int> perm) {
-          std::sort(perm.begin(), perm.end());
-          for (int i = 0; i < static_cast<int>(perm.size()); ++i) {
-            if (perm[i] != i) {
-              return false;
-            }
+  explicit inline Transpose(const OpSpec &spec)
+      : Operator<Backend>(spec)
+      , perm_(spec.GetRepeatedArgument<int>("perm"))
+      , transpose_layout_(spec.GetArgument<bool>("transpose_layout"))
+      , output_layout_arg_(spec.GetArgument<TensorLayout>("output_layout")) {
+    DALI_ENFORCE(
+      [](std::vector<int> perm) {
+        std::sort(perm.begin(), perm.end());
+        for (int i = 0; i < static_cast<int>(perm.size()); ++i) {
+          if (perm[i] != i) {
+            return false;
           }
-          return true;
-        }(perm_), "Invalid permutation: sorted `perm` is not equal to [0, ..., n-1].");
+        }
+        return true;
+      }(perm_), "Invalid permutation: sorted `perm` is not equal to [0, ..., n-1].");
     }
 
   ~Transpose() override;
 
   DISABLE_COPY_MOVE_ASSIGN(Transpose);
 
  protected:
-  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+  bool SetupImpl(std::vector<OutputDesc> &output_desc,
+                 const workspace_t<Backend> &ws) override {
+    const auto &input = ws.template Input<Backend>(0);
+    auto in_layout = input.GetLayout();
+    output_layout_ = in_layout;
+    if (!output_layout_arg_.empty()) {
+      output_layout_ = output_layout_arg_;
```

Comment:
It'd be nice to enforce that `output_layout_.ndim() == input.shape().sample_dim()`.

Reply:
will do